Watchman.Discord/Areas/Protection/Services/MuteService.cs

```diff
@@ -49,42 +50,55 @@ public async Task MuteUserOrOverwrite(Contexts contexts, MuteEvent muteEvent, Us
 
             var messagesService = this._messagesServiceFactory.Create(contexts);
             await messagesService.SendResponse(x => x.MutedUser(userToMute, muteEvent.TimeRange.End));
+            await this._directMessagesService.TrySendMessage(userToMute.Id, x => x.YouHaveBeenMuted(userToMute, muteEvent.TimeRange.End, muteEvent.Reason), contexts);
         }
 
         public async Task<bool> UnmuteIfNeeded(DiscordServerContext server, UserContext userToUnmute)
         {
-            var userMuteEvents = this.GetMuteEvents(server, userToUnmute.Id);
-            var eventToUnmute = this.GetNotUnmutedEvent(userMuteEvents);
+            var userMuteEvents = this.GetUserMuteEvents(server, userToUnmute.Id);
+            var eventToUnmute = userMuteEvents.FirstOrDefault(x => x.Unmuted == false);
             if (eventToUnmute == null)
             {
                 return false;
             }
-            userToUnmute = this._usersService.GetUserById(server, userToUnmute.Id);
-            if (userToUnmute == null) // user could left the server
-            {
-                return false;
-            }
-            await this.UnmuteUser(userToUnmute, eventToUnmute, server);
-            return true;
+            return await this.UnmuteIfNeededSpecificEvent(server, userToUnmute, eventToUnmute);
         }
 
-        public async void UnmuteInFuture(Contexts contexts, MuteEvent muteEvent, UserContext userToUnmute)
+        public async void UnmuteInFuture(Contexts contexts, MuteEvent muteEvent, UserContext userToUnmute, bool sendMessageAfterUnmute = true)
         {
-            var timeRange = muteEvent.TimeRange;
-            await Task.Delay(timeRange.End - timeRange.Start);
-
-            var wasNeededToUnmute = await this.UnmuteIfNeeded(contexts.Server, userToUnmute);
-            if (wasNeededToUnmute)
+            if (muteEvent.TimeRange.End > DateTime.UtcNow)
+            {
+                await Task.Delay(muteEvent.TimeRange.End - DateTime.UtcNow);
```

Comment:
pytanie czy jak byłoby dużo serwerów i dużo mutów
to byśmy sobie z tym poradzili wydajnościowo 

może warto użyć jakiejś lokalnej bazy danych która byłaby bardzo szybka
i odpytywać ją powiedzmy co 15 minut 

i w taki sposób jak tutaj obsługiwać tylko te, którym brakuje mniej niż 15 minut 

wiesz, żeby nie było tak że trzymamy bardzo dużo mutów w pamięci
z czego większość to są jakieś kilkudniowe 

na naszym serwerze tego nie ma
ale warto o tym myśleć przyszłościowo 

nie mówie żeby przy tym zadaniu to przebudowywać, ale warto się nad tym zastanowić

Reply:
też o tym myślę
można zrobić nowy serwis który będzie oparty na hangfire i będzie to ogarniał

Reply:
dobry pomysł